utils/server.js

```diff
@@ -0,0 +1,168 @@
+        !customOptions.enableBackgroundAutoReload &&
+        !customOptions.enableContentScriptsAutoReload
+      ) {
+        return middlewares;
+      }
+
+      if (!devServer) {
+        throw new Error('webpack-dev-server is not defined');
+      }
+
+      // imagine you are using app.use(path, middleware) in express.
+      // in fact, devServer is an express server.
+      middlewares.push({
+        path: '/__server_sent_events__', // you can find this path requested by backgroundClient.js.
+        middleware: (req, res) => {
+          const sseStream = new SSEStream(req);
+          sseStream.pipe(res);
+
+          sseStream.write('message from webserver.');
+
+          let closed = false;
+
+          const compileDoneHook = debounce((stats) => {
+            const { modules } = stats.toJson({ all: false, modules: true });
+            const updatedJsModules = modules.filter(
+              (module) =>
+                module.type === 'module' &&
+                module.moduleType === 'javascript/auto'
+            );
+
+            const isBackgroundUpdated = updatedJsModules.some((module) =>
+              module.nameForCondition.startsWith(
+                path.resolve(__dirname, '../src/pages/Background')
+              )
+            );
+            const isContentScriptsUpdated = updatedJsModules.some((module) =>
+              module.nameForCondition.startsWith(
+                path.resolve(__dirname, '../src/pages/ContentScripts')
+              )
+            );
+
+            const shouldBackgroundReload =
+              !stats.hasErrors() &&
+              isBackgroundUpdated &&
+              customOptions.enableBackgroundAutoReload;
+            const shouldContentScriptsReload =
+              !stats.hasErrors() &&
+              isContentScriptsUpdated &&
+              customOptions.enableContentScriptsAutoReload;
+
+            if (shouldBackgroundReload) {
+              sseStream.writeMessage(
+                {
+                  event: 'background-updated',
+                  data: {}, // "data" key should be reserved though it is empty.
+                },
+                'utf-8'
+              );
+            }
+            if (shouldContentScriptsReload) {
+              sseStream.writeMessage(
+                {
+                  event: 'content-scripts-updated',
+                  data: {},
+                },
+                'utf-8'
+              );
+            }
+          }, 1000);
+
+          const plugin = (stats) => {
+            if (!closed) {
+              compileDoneHook(stats);
+            }
+          };
+
+          // a mini webpack plugin just born!
+          // this plugin will be triggered after each compilation done.
+          compiler.hooks.done.tap('extension-auto-reload-plugin', plugin);
+
+          res.on('close', () => {
+            closed = true;
+            sseStream.unpipe(res);
+          });
+        },
+      });
+
+      return middlewares;
+    },
```

Comment:
Add a middleware to webpack devServer (which basicly is a Express server).

This middleware handles the EventSrouce request from backgroundClient, and start a long connection. There is also a mini webpack plugin implemented here to watch the webpack compilation event so each time Background or ContentScripts related code is changed, the devServer can send events to backgroundClient.

Then backgroundClient makes auto-reload happen in browser.